Apps/FlightSoftware/Watchdog/src/i2c_sensors.c

```diff
@@ -0,0 +1,575 @@
+                internals.gState = nextState;
+                *((uint8_t*) output) = tStatus.data;
+                *done = TRUE;
+                *gotOutput = TRUE;
+                continueSpinning = TRUE;
+            }
+
+            prevReadDone = TRUE;
+        } else if (I2C__TRANSACTION__DONE_ERROR_NACK == tStatus.state) {
+            // We're now done with this read (and it failed)
+            if (statusForRequestedTransaction) {
+                internals.readings.nackMask |= nackMaskBit;
+                internals.gState = nextState;
+                *done = TRUE;
+                *gotOutput = FALSE;
+                continueSpinning = TRUE;
+            }
+
+            prevReadDone = TRUE;
+        }
+
+        startNewRead = (!statusForRequestedTransaction && prevReadDone);
+    }
+
+    if (startNewRead) {
+        // If the NACK mask bit for this register is already set, just skip
+        // this register
+        if (internals.readings.nackMask & nackMaskBit != 0) {
+            internals.gState = nextState;
+            *done = TRUE;
+            *gotOutput = FALSE;
+            return TRUE;
+        }
+
+        stat = I2C__read(devAddr, regAddr);
+
+        assert(I2C__STATUS__SUCCESS == stat);
+
+        *done = FALSE;
+        *gotOutput = FALSE;
+    }
+
+
+    return continueSpinning;       
+}
+
+void I2C_Sensors__writeRegNonBlocking(uint8_t devAddr,
+                                      uint8_t regAddr,
+                                      uint8_t data,
+                                      BOOL* done,
+                                      BOOL* success)
+{
+    static I2C__TransactionStatus tStatus = { 0 };
+
+    I2C__Status stat = I2C__getTransactionStatus(&tStatus);
+
+    BOOL startNewWrite = FALSE;
+
+    if (I2C__STATUS__ERROR__NO_TRANSACTION == stat) {
+        startNewWrite = TRUE;
+    } else if (I2C__STATUS__SUCCESS == stat) {
+        BOOL statusForRequestedTransaction = FALSE;
+        BOOL prevWriteDone = FALSE;
+
+        if ((tStatus.devAddr == devAddr) && (tStatus.regAddr == regAddr)) {
+            // This transaction status is for the currently requested write
+            statusForRequestedTransaction = TRUE;
+        }
+        
+        if (I2C__TRANSACTION__DONE_SUCCESS == tStatus.state) {
+            // We're now done with this write (and it was successful)
+            if (statusForRequestedTransaction) {
+                *done = TRUE;
+                *success = TRUE;
+            }
+
+            prevWriteDone = TRUE;
+        } else if (I2C__TRANSACTION__DONE_ERROR_NACK == tStatus.state) {
+            // We're now done with this write (and it failed)
+            if (statusForRequestedTransaction) {
+                *done = TRUE;
+                *success = FALSE;
+            }
+
+            prevWriteDone = TRUE;
+        }
+
+        startNewWrite = (!statusForRequestedTransaction && prevWriteDone);
+    }
+
+    if (startNewRead) {
```

Comment:
Ah okay, so yeah I was definitely calling it incorrectly, you can see where I was calling it [here](https://github.com/PlanetaryRobotics/CubeRoverPackage/blob/7c93d7df7c9998504453119a2ef60d8b0cb947b1/Apps/FlightSoftware/Watchdog/src/main.c#L214). 

Also just as an FYI I am out of town until Monday morning, so I won’t be able to test anything more until then. I do think your input should be enough to get it all working properly though, will keep you posted.

Thanks for the help!

Reply:
Also, could it be that because I was not waiting for the i2c transaction to complete, the sensor (slave) is trying to do something else which is why the master is stuck forever trying trying to do `txStart` but the sensor will never ACK because it's waiting for something else? Just a thought

Reply:
Is there a reason you're not just using [my implementation of how to call it](https://github.com/PlanetaryRobotics/CubeRoverPackage/blob/e9078468f3c34df989645fac35748a36f0d72874/Apps/FlightSoftware/Watchdog/src/main.c#L209-L239)? Also please please please check return values. It doesn't appear a way has been set up to log debug or error messages on the watchdog, which is significantly less than ideal, but you should still be able to make asserts and other stuff that will help you debug. @ohnx, is @SchwartzCode the only one that can test this out this weeked? I'm concerned about putting things off to the last minute, and I'm also traveling for work next week (yes, again >:( ) so I'll have less time and energy to respond than I will over this weekend.

Reply:
In response to your theory about what we're seeing, I guess in theory it's possible. However, `I2C__read()` should just be returning an error if the previous transaction isn't complete. That being said, I believe that [this issue](https://github.com/PlanetaryRobotics/CubeRoverPackage/pull/59#discussion_r632537439) would have caused the gauge readings to never complete. I'm a little curious if perhaps we weren't getting stuck in txStart, but maybe instead it just takes longer than the other states and because we were infinitely looping in I2C_Sensors we were always just reading the same device and register over and over and over -- make that looked like we were getting stuck in txStart when that wasn't the case? Either way, what I really want to see is how it behaves if you make all the changes I've commented in this review and then re-test using my implementation of how to call into I2C_Sensors.